File: tool/src/main/java/org/metalscraps/eso/lang/tool/Destinations.java
package org.metalscraps.eso.lang.tool;

import lombok.AllArgsConstructor;
import org.apache.commons.io.FileUtils;
import org.metalscraps.eso.lang.lib.bean.PO;
import org.metalscraps.eso.lang.lib.config.AppConfig;
import org.metalscraps.eso.lang.lib.config.AppWorkConfig;
import org.metalscraps.eso.lang.lib.config.FileNames;
import org.metalscraps.eso.lang.lib.config.SourceToMapConfig;
import org.metalscraps.eso.lang.lib.util.Utils;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by 안병길 on 2018-01-15.
 * Whya5448@gmail.com
 */


@AllArgsConstructor
class Destinations {

	private final AppWorkConfig appWorkConfig;

	private void work(FileNames[] fileNames, File target){

		HashMap<String, PO> poMap = new HashMap<>();

		try {
			StringBuilder destinationQuestSource = new StringBuilder(FileUtils.readFileToString(new File(target.getPath().replace("kr","en")), AppConfig.CHARSET));
			SourceToMapConfig config = new SourceToMapConfig().setPattern(AppConfig.POPattern).setKeyGroup(6).setPrefix("{\"").setSuffix("\"}").setProcessItemName(false);

			for(FileNames fileName : fileNames) poMap.putAll(Utils.sourceToMap(config.setFile(new File(appWorkConfig.getPODirectory() +"/"+ fileName.toStringPO2()))));

			boolean init = false;
			for(Map.Entry<String, PO> entry : new HashMap<>(poMap).entrySet()) {
				if(!init) { poMap.clear(); init = true; }
				PO po = entry.getValue();
				if(po.isFuzzy()) po.setTarget(po.getSource());
				poMap.put("{\""+entry.getKey()+"\"}", po);
			}

			// 찾아바꾸기
			Utils.replaceStringFromMap(destinationQuestSource, poMap);

			// 저장
			System.out.println(target.getAbsolutePath());
			FileUtils.write(new File(target.getAbsolutePath()), destinationQuestSource, AppConfig.CHARSET);

		} catch(Exception e) {
			e.printStackTrace();
		}
	}

	void start() {

		work(new FileNames[] { FileNames.journey, FileNames.journeyOther } , new File(appWorkConfig.getBaseDirectory()+"/Destinations/DestinationsQuests_kr.lua"));
		work(new FileNames[] { FileNames.npcName } , new File(appWorkConfig.getBaseDirectory()+"/Destinations/DestinationsQuestgivers_kr.lua"));

	}
}

File: tool/src/main/java/org/metalscraps/eso/lang/tool/Destinations.kt
package org.metalscraps.eso.lang.tool

import org.apache.commons.io.FileUtils
import org.metalscraps.eso.lang.lib.bean.PO
import org.metalscraps.eso.lang.lib.config.AppConfig
import org.metalscraps.eso.lang.lib.config.AppWorkConfig
import org.metalscraps.eso.lang.lib.config.FileNames
import org.metalscraps.eso.lang.lib.config.SourceToMapConfig
import org.metalscraps.eso.lang.lib.util.Utils
import org.slf4j.LoggerFactory

import java.nio.file.Path
import java.util.HashMap

/**
 * Created by 안병길 on 2018-01-15.
 * Whya5448@gmail.com
 */


internal class Destinations(private val config: AppWorkConfig) {
    private val logger = LoggerFactory.getLogger(Destinations::class.java)

    private fun work(fileNames: Array<FileNames>, target: Path) {

        val poMap = HashMap<String, PO>()

        try {
            val destinationQuestSource = StringBuilder(FileUtils.readFileToString(target.resolveSibling(target.fileName.toString().replace("kr", "en")).toFile(), AppConfig.CHARSET))
            val config = SourceToMapConfig().setPattern(AppConfig.POPattern).setKeyGroup(6).setPrefix("{\"").setSuffix("\"}").setProcessItemName(false)

            for (fileName in fileNames) poMap.putAll(Utils.sourceToMap(config.setPath(this.config.poDirectoryToPath.resolve(fileName.toStringPO2()))))

            var init = false
            for ((key, po) in HashMap(poMap)) {
                if (!init) {
                    poMap.clear()
                    init = true
                }
                if (po.isFuzzy) po.target = po.source
                poMap["{\"$key\"}"] = po
            }

            // 찾아바꾸기
            Utils.replaceStringFromMap(destinationQuestSource, poMap)

            // 저장
            logger.info("" + target)
            FileUtils.write(target.toFile(), destinationQuestSource, AppConfig.CHARSET)

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun start() {

        work(arrayOf(FileNames.journey, FileNames.journeyOther), config.baseDirectoryToPath.resolve("Destinations/DestinationsQuests_kr.lua"))
        work(arrayOf(FileNames.npcName), config.baseDirectoryToPath.resolve("Destinations/DestinationsQuestgivers_kr.lua"))

    }

}
